### ISSUE TITLE
List comprehension with conditional still triggers error for Optional

### ISSUE THREAD

**Issue by nkakouros** [python/mypy#8596]
* Are you reporting a bug, or opening a feature request?
Bug

* Please insert below the code you are checking with mypy
```python
#!/usr/bin/env python

from typing import Sequence


class A:
    def __init__(self, a: Sequence = None):
        self.a = a


a = A([1, 2, 3, 4])

[_ for _ in a.a if a.a]

if a.a:
    for _ in a.a:
        pass
```

* What is the actual behavior/output?

The list comprehension complains that `Item "None" of "Optional[Sequence[Any]]" has no attribute "__iter__" (not iterable)`. The loop that follows however does not trigger the error.

* What is the behavior/output you expect?
I would expect the comprehension to not trigger an error.

* What are the versions of mypy and Python you are using?

```
mypy 0.770+dev.52c0a63f4604a2bfa8c0edd6589ab25fab4bdaec (git master as of today)
python 3.8.1
```

**Comment by ilevkivskyi** [comment]
Hm, interesting, this is indeed a bug. Looks like a missing corner case for binder.

**Comment by JelleZijlstra** [comment]
This is not a bug in mypy, because the list comprehension would always try to iterate over `a.a`. The listcomp code is equivalent to:
```
for _ in a.a:
    if a.a:
       _
```
Mypy correctly identified an error in the code.

**Comment by ilevkivskyi** [comment]
Yeah, this is why I shouldn't triage issues so late.

**Comment by nkakouros** [comment]
Thanks for the explanation.
